### PR TITLE
IGNITE-13888 Provide the utility to output performance statistics operations to the console

### PR DESCRIPTION
**Pull request by NSAmelchev**
The utility to print performance statistics operations to the console or file

### CODE REVIEW

File: modules/performance-statistics-ext/src/main/java/org/apache/ignite/internal/performancestatistics/PerformanceStatisticsPrinter.java [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+        if (params.outFile != null) {
+            try {
+                ps = new PrintStream(new BufferedOutputStream(new FileOutputStream(new File(params.outFile))));
+            }
+            catch (IOException e) {
+                throw new IllegalArgumentException("Cannot write to output file", e);
+            }
+        }
+        else
+            ps = System.out;
```
Comment (nizhikov) [inline modules/performance-statistics-ext/src/main/java/org/apache/ignite/internal/performancestatistics/PerformanceStatisticsPrinter.java:64]: Let's extract this to the separate method.

File: modules/performance-statistics-ext/src/main/java/org/apache/ignite/internal/performancestatistics/PerformanceStatisticsPrinter.java [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+
+        try {
+            PrintHandler hnd = new PrintHandler(ps, params.ops, params.from, params.to, params.cacheIds);
```
Comment (nizhikov) [inline modules/performance-statistics-ext/src/main/java/org/apache/ignite/internal/performancestatistics/PerformanceStatisticsPrinter.java:67]: `hnd` variable can be inlined.

Comment (NSAmelchev) [reply modules/performance-statistics-ext/src/main/java/org/apache/ignite/internal/performancestatistics/PerformanceStatisticsPrinter.java]: Done.

File: modules/performance-statistics-ext/src/main/java/org/apache/ignite/internal/performancestatistics/PerformanceStatisticsPrinter.java (hunk 1 above)
Comment (NSAmelchev) [reply modules/performance-statistics-ext/src/main/java/org/apache/ignite/internal/performancestatistics/PerformanceStatisticsPrinter.java]: Done.

File: modules/performance-statistics-ext/src/main/java/org/apache/ignite/internal/performancestatistics/PerformanceStatisticsPrinter.java [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+            U.nl() +
+            "Times must be specified in the Unix time format in milliseconds." + U.nl() +
+            "List of operation types: " + ops + '.');
+
+        System.exit(0);
+    }
+
+    /** @param params Validates parameters. */
+    private static void validateParameters(Parameters params) {
+        File statFiles = new File(params.statFileOrDir);
```
Comment (nizhikov) [inline modules/performance-statistics-ext/src/main/java/org/apache/ignite/internal/performancestatistics/PerformanceStatisticsPrinter.java:157]: statFiles -> statFileOrDir

File: modules/performance-statistics-ext/src/main/java/org/apache/ignite/internal/performancestatistics/PerformanceStatisticsPrinter.java [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+
+        A.ensure(statFiles.exists(), "Performance statistics file or files directory does not exists");
+
+        if (params.outFile != null) {
+            File out = new File(params.outFile);
+
+            try {
+                boolean created = out.createNewFile();
+
+                if (!created) {
```
Comment (nizhikov) [inline modules/performance-statistics-ext/src/main/java/org/apache/ignite/internal/performancestatistics/PerformanceStatisticsPrinter.java:167]: Can we just append to the existing file?

File: modules/performance-statistics-ext/src/main/java/org/apache/ignite/internal/performancestatistics/PerformanceStatisticsPrinter.java (hunk 3 above)
Comment (NSAmelchev) [reply modules/performance-statistics-ext/src/main/java/org/apache/ignite/internal/performancestatistics/PerformanceStatisticsPrinter.java]: Fixed.

File: modules/performance-statistics-ext/src/main/java/org/apache/ignite/internal/performancestatistics/PerformanceStatisticsPrinter.java (hunk 4 above)
Comment (NSAmelchev) [reply modules/performance-statistics-ext/src/main/java/org/apache/ignite/internal/performancestatistics/PerformanceStatisticsPrinter.java]: Done.

File: modules/performance-statistics-ext/src/main/java/org/apache/ignite/internal/performancestatistics/handlers/PrintHandler.java [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+ */
+public class PrintHandler implements PerformanceStatisticsHandler {
+    /** Json mapper. */
+    private static final ObjectMapper MAPPER = new ObjectMapper();
+
+    /** Print stream. */
+    private final PrintStream ps;
+
+    /** */
+    StringBuilder sb = new StringBuilder();
```
Comment (nizhikov) [inline modules/performance-statistics-ext/src/main/java/org/apache/ignite/internal/performancestatistics/handlers/PrintHandler.java:53]: Let's get rid of `sb` and append all changes directly to the print stream.

Comment (NSAmelchev) [reply modules/performance-statistics-ext/src/main/java/org/apache/ignite/internal/performancestatistics/handlers/PrintHandler.java]: Done.

File: modules/performance-statistics-ext/src/main/java/org/apache/ignite/internal/performancestatistics/handlers/PrintHandler.java [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+        if (skip(CACHE_START, cacheId))
+            return;
+
+        ps.print("{\"op\":\"" + CACHE_START);
+        ps.print("\",\"nodeId\":\"");
+        ps.print(nodeId);
+        ps.print("\",\"cacheId\":");
+        ps.print(cacheId);
+        ps.print(",\"name\":");
+        ps.print(new TextNode(name));
```
Comment (nizhikov) [inline modules/performance-statistics-ext/src/main/java/org/apache/ignite/internal/performancestatistics/handlers/PrintHandler.java:86]: We should avoid the creation of intermediate objects just to create an escaped strings.

Comment (NSAmelchev) [reply modules/performance-statistics-ext/src/main/java/org/apache/ignite/internal/performancestatistics/handlers/PrintHandler.java]: Done.

File: modules/performance-statistics-ext/src/main/java/org/apache/ignite/internal/performancestatistics/util/Utils.java [hunk 7]
```diff
@@ -51,4 +56,43 @@ public static ArrayNode createArrayIfAbsent(String val, ObjectNode json) {
 
         return node;
     }
+
+    /**
+     * Prints JSON-escaped string to the stream.
+     *
+     * @param ps Print stream to write to.
+     * @param str String to print.
```
Comment (nizhikov) [inline modules/performance-statistics-ext/src/main/java/org/apache/ignite/internal/performancestatistics/util/Utils.java:64]: Please, add @see annotation to the base method for this.

**Review by nizhikov — APPROVED** [review]

**Pull request merged**